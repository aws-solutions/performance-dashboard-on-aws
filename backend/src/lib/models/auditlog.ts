/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
