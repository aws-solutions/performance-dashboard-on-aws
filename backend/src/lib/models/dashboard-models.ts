export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};