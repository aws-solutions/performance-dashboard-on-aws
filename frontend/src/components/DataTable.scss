@import "../styles/base";

[data-reach-menu-item][data-selected] a {
    color: color("white") !important;
}
