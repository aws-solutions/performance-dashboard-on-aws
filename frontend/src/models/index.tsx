/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
