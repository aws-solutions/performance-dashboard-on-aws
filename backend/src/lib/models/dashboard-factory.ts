import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
