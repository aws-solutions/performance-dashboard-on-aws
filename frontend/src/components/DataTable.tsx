/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, useMemo } from "react";
import { ColumnMetadata } from "../models";
import { useTranslation } from "react-i18next";
import UtilsService from "../services/UtilsService";
import TickFormatter from "../services/TickFormatter";
import Table from "./Table";
import DropdownMenu from "../components/DropdownMenu";
import { CSVLink } from "react-csv";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faDownload, faEye, faEyeSlash } from "@fortawesome/free-solid-svg-icons";
import "./DataTable.scss";
import { useHistory } from "react-router-dom";

const { MenuItem } = DropdownMenu;

interface Props {
    rows: any[];
    columns: string[];
    columnsMetadata?: ColumnMetadata[];
    fileName?: string;
    showMobilePreview?: boolean;
    title: string;
}

function DataTable({ rows, columns, columnsMetadata, fileName, showMobilePreview, title }: Props) {
    const history = useHistory();
    const { t } = useTranslation();
    const [showDataTable, setShowDataTable] = useState(false);
    let canDownload = false;

    const tableRows = useMemo(() => rows, [rows]);
    const tableColumns = useMemo(
        () =>
            columns.map((header) => {
                return {
                    Header: header,
                    id: header,
                    accessor: header,
                    minWidth: 150,
                    Cell: (props: any) => {
                        const row = props.row.original;

                        // Check if there is metadata for this column
                        let columnMetadata;
                        if (columnsMetadata) {
                            columnMetadata = columnsMetadata.find((cm) => cm.columnName === header);
                        }

                        return !UtilsService.isCellEmpty(row[header])
                            ? TickFormatter.format(row[header], 0, false, "", "", columnMetadata)
                            : "-";
                    },
                };
            }),
        [columns, columnsMetadata],
    );

    return (
        <>
            <div className={showMobilePreview ? "margin-left-05" : "text-right margin-right-05"}>
                <DropdownMenu
                    className="text-base"
                    buttonText={t("Actions")}
                    disabled={false}
                    variant="unstyled"
                    ariaLabel={`${title} - ${t("ARIA.DataTableActions")}`}
                >
                    <MenuItem
                        onSelect={() =>
                            !showDataTable ? setShowDataTable(true) : setShowDataTable(false)
                        }
                        aria-label={
                            !showDataTable ? t("ShowDataTableButton") : t("HideDataTableButton")
                        }
                    >
                        <FontAwesomeIcon
                            icon={!showDataTable ? faEye : faEyeSlash}
                            className="margin-right-1 margin-bottom-1px"
                            size="xs"
                        />
                        {!showDataTable ? t("ShowDataTableButton") : t("HideDataTableButton")}
                    </MenuItem>
                    <MenuItem
                        onSelect={() => {
                            const downloadButton: HTMLAnchorElement | null = document.querySelector(
                                "[data-reach-menu-item][data-selected] a",
                            );
                            if (downloadButton) {
                                // de-duplicate clicks when link clicked
                                canDownload = true;
                                downloadButton.click();
                                canDownload = false;
                            }
                        }}
                        aria-label={t("DownloadCSV")}
                    >
                        <FontAwesomeIcon icon={faDownload} className="margin-right-1" size="xs" />
                        <CSVLink
                            data={tableRows}
                            filename={`${fileName}.csv`}
                            className="usa-link"
                            onClick={() => {
                                if (!canDownload) {
                                    return false;
                                }
                                history.replace(history.location.pathname, {
                                    alert: {
                                        type: "success",
                                        message: t("DownloadFileSuccess", {
                                            fileName: `${fileName}.csv`,
                                        }),
                                    },
                                });
                                return true;
                            }}
                        >
                            {t("DownloadCSV")}
                        </CSVLink>
                    </MenuItem>
                </DropdownMenu>
            </div>
            {showDataTable && (
                <Table
                    id="items"
                    selection="none"
                    rows={tableRows}
                    disablePagination={true}
                    columns={tableColumns}
                    pageSize={50}
                />
            )}
        </>
    );
}

export default DataTable;
