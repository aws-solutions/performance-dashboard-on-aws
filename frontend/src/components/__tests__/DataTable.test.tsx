/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { ColumnDataType, ColumnMetadata, CurrencyDataType, NumberDataType } from "../../models";
import { render, screen } from "@testing-library/react";
import DataTable from "../DataTable";
import userEvent from "@testing-library/user-event";

const columns = ["year", "sales", "revenue"];
const rows = [
    {
        year: 2015,
        sales: 10000,
        revenue: 34,
    },
    {
        year: 2016,
        sales: 45000,
        revenue: 74,
    },
    {
        year: 2017,
        sales: 56000,
        revenue: 23,
    },
];

const columnsMetadata: ColumnMetadata[] = [
    {
        columnName: "year",
        dataType: ColumnDataType.Text,
        hidden: false,
    },
    {
        columnName: "sales",
        dataType: ColumnDataType.Number,
        numberType: NumberDataType.Currency,
        currencyType: CurrencyDataType["Dollar $"],
        hidden: false,
    },
    {
        columnName: "revenue",
        dataType: ColumnDataType.Number,
        numberType: NumberDataType.Percentage,
        hidden: false,
    },
];

test("table should be hidden by default", async () => {
    render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={false}
            title={"test-title"}
        />,
    );
    expect(screen.queryByRole("table")).not.toBeInTheDocument();
});

test("shows and hides table when button is clicked", async () => {
    render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={false}
            title={"test-title"}
        />,
    );
    const showTableBtn = screen.getByText("Show data table");
    userEvent.click(showTableBtn);
    expect(screen.getByRole("table")).toBeInTheDocument();

    const hideTableBtn = screen.getByText("Hide data table");
    userEvent.click(hideTableBtn);
    expect(screen.queryByRole("table")).not.toBeInTheDocument();
});

test("displays a table with values properly formatted", async () => {
    render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={false}
            title={"test-title"}
        />,
    );
    const showTableBtn = screen.getByText("Show data table");
    userEvent.click(showTableBtn);

    // Years should be formatted as text as per ColumnMetadata, without comma separators
    expect(screen.getByText("2015")).toBeInTheDocument();
    expect(screen.getByText("2016")).toBeInTheDocument();
    expect(screen.getByText("2017")).toBeInTheDocument();

    // Sales column should be formatted as currency
    expect(screen.getByText("$10,000")).toBeInTheDocument();
    expect(screen.getByText("$45,000")).toBeInTheDocument();
    expect(screen.getByText("$56,000")).toBeInTheDocument();

    // Revenue column should be formatted as percentage
    expect(screen.getByText("34%")).toBeInTheDocument();
    expect(screen.getByText("74%")).toBeInTheDocument();
    expect(screen.getByText("23%")).toBeInTheDocument();
});

test("renders a Data Table element with hidden table", async () => {
    const wrapper = render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={false}
            title={"test-title"}
        />,
    );

    expect(wrapper.container).toMatchSnapshot();
});

test("renders a Data Table element with hidden table on mobile", async () => {
    const wrapper = render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={true}
            title={"test-title"}
        />,
    );

    expect(wrapper.container).toMatchSnapshot();
});

test("renders a Data Table element", async () => {
    const wrapper = render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={false}
            title={"test-title"}
        />,
    );

    const showTableBtn = wrapper.getByText("Show data table");
    userEvent.click(showTableBtn);

    expect(wrapper.container).toMatchSnapshot();
});

test("renders a Data Table element on mobile", async () => {
    const wrapper = render(
        <DataTable
            rows={rows}
            columns={columns}
            columnsMetadata={columnsMetadata}
            fileName={"test-file-name"}
            showMobilePreview={true}
            title={"test-title"}
        />,
    );

    const showTableBtn = wrapper.getByText("Show data table");
    userEvent.click(showTableBtn);

    expect(wrapper.container).toMatchSnapshot();
});
