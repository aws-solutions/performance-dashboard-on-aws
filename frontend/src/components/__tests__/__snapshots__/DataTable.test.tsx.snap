// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Data Table element 1`] = `
<div>
  <div
    class="text-right margin-right-05"
  >
    <button
      aria-controls="menu--33"
      aria-haspopup="true"
      aria-label="test-title - Data table actions"
      class="usa-button usa-button--unstyled text-base"
      data-reach-menu-button=""
      id="menu-button--menu--33"
      type="button"
    >
      Actions
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
  <div
    class="overflow-x-hidden overflow-y-hidden"
  >
    <table
      class="usa-table usa-table--borderless"
      id="items"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: year"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: year"
              type="button"
            >
              <span>
                year
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: sales"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: sales"
              type="button"
            >
              <span>
                sales
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: revenue"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: revenue"
              type="button"
            >
              <span>
                revenue
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2015
          </th>
          <td
            role="cell"
          >
            $10,000
          </td>
          <td
            role="cell"
          >
            34%
          </td>
        </tr>
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2016
          </th>
          <td
            role="cell"
          >
            $45,000
          </td>
          <td
            role="cell"
          >
            74%
          </td>
        </tr>
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2017
          </th>
          <td
            role="cell"
          >
            $56,000
          </td>
          <td
            role="cell"
          >
            23%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders a Data Table element on mobile 1`] = `
<div>
  <div
    class="margin-left-05"
  >
    <button
      aria-controls="menu--40"
      aria-haspopup="true"
      aria-label="test-title - Data table actions"
      class="usa-button usa-button--unstyled text-base"
      data-reach-menu-button=""
      id="menu-button--menu--40"
      type="button"
    >
      Actions
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
  <div
    class="overflow-x-hidden overflow-y-hidden"
  >
    <table
      class="usa-table usa-table--borderless"
      id="items"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: year"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: year"
              type="button"
            >
              <span>
                year
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: sales"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: sales"
              type="button"
            >
              <span>
                sales
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
          <th
            aria-sort="none"
            colspan="1"
            role="columnheader"
            scope="col"
            style="min-width: 150px;"
          >
            <button
              aria-label="Toggle SortBy: revenue"
              class="usa-button usa-button--unstyled text-no-underline"
              style="cursor: pointer;"
              title="Toggle SortBy: revenue"
              type="button"
            >
              <span>
                revenue
              </span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                data-icon="chevron-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2015
          </th>
          <td
            role="cell"
          >
            $10,000
          </td>
          <td
            role="cell"
          >
            34%
          </td>
        </tr>
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2016
          </th>
          <td
            role="cell"
          >
            $45,000
          </td>
          <td
            role="cell"
          >
            74%
          </td>
        </tr>
        <tr
          role="row"
        >
          <th
            role="cell"
            scope="row"
          >
            2017
          </th>
          <td
            role="cell"
          >
            $56,000
          </td>
          <td
            role="cell"
          >
            23%
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders a Data Table element with hidden table 1`] = `
<div>
  <div
    class="text-right margin-right-05"
  >
    <button
      aria-controls="menu--23"
      aria-haspopup="true"
      aria-label="test-title - Data table actions"
      class="usa-button usa-button--unstyled text-base"
      data-reach-menu-button=""
      id="menu-button--menu--23"
      type="button"
    >
      Actions
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`renders a Data Table element with hidden table on mobile 1`] = `
<div>
  <div
    class="margin-left-05"
  >
    <button
      aria-controls="menu--28"
      aria-haspopup="true"
      aria-label="test-title - Data table actions"
      class="usa-button usa-button--unstyled text-base"
      data-reach-menu-button=""
      id="menu-button--menu--28"
      type="button"
    >
      Actions
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 margin-left-1"
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
